### PR TITLE
new usage of amp training.

### PR DESCRIPTION
**Pull request by huangxu96**
混合精度训练-纯fp16训练的更简洁用法。

执行脚本参考：

```

export CUDA_VISIBLE_DEVICES=5
export FLAGS_fraction_of_gpu_memory_to_use=0.8

batch_size=208
python3 static/train.py \
        -c ../configs/ResNet/ResNet50_fp16.yaml \
        -o TRAIN.batch_size=${batch_size} \
        -o validate=False \
        -o epochs=1 \
        -o TRAIN.data_dir=/datasets/ILSVRC2012 \
        -o TRAIN.file_list=/datasets/ILSVRC2012/train_list.txt \
        -o TRAIN.num_workers=4 

```

### CODE REVIEW

File: ppcls/modeling/loss.py [hunk 1]
```diff
@@ -42,17 +42,14 @@ def _labelsmoothing(self, target):
         soft_target = paddle.reshape(soft_target, shape=[-1, self._class_dim])
         return soft_target
```
Comment (littletomatodonkey) [inline ppcls/modeling/loss.py:44]: 这个文件需要恢复一下吧，不然之前都是求mean的

File: ppcls/modeling/architectures/resnet.py [hunk 2]
```diff
@@ -277,14 +277,18 @@ def __init__(self, layers=50, class_dim=1000, input_image_channel=3, data_format
...
     def forward(self, inputs):
-        y = self.conv(inputs)
-        y = self.pool2d_max(y)
-        for block in self.block_list:
-            y = block(y)
-        y = self.pool2d_avg(y)
-        y = paddle.reshape(y, shape=[-1, self.pool2d_avg_channels])
-        y = self.out(y)
-        return y
+        with paddle.static.amp.fp16_guard():
```
Comment (littletomatodonkey) [inline ppcls/modeling/architectures/resnet.py:280]: 加一个判断吧，在像下面这种，传一个参数进来，保持默认情况下和之前的运行模式相同，

image

File: tools/static/program.py [hunk 3]
```diff
@@ -21,12 +21,10 @@
 import numpy as np
 
 from collections import OrderedDict
-from optimizer import OptimizerBuilder
+from .optimizer import OptimizerBuilder
```
Comment (littletomatodonkey) [inline tools/static/program.py:24]: 这块是不是就可以和ppcls/optimizer/optimizer.py统一写法了，要不统一一下写法吧，去除static下的这个optimizer

File: tools/static/program.py [hunk 4]
```diff
@@ -239,9 +225,8 @@ def create_fetchs(out,
         fetchs(dict): dict of model outputs(included loss and measures)
     """
     fetchs = OrderedDict()
-    use_pure_fp16 = config.get("use_pure_fp16", False)
     loss = create_loss(out, feeds, architecture, classes_num, epsilon, use_mix,
-                       use_distillation, use_pure_fp16)
+                       use_distillation)
```
Comment (littletomatodonkey) [inline tools/static/program.py:229]: 同上，要提供一个参数是做paddle.mean和paddle.sum

File: tools/static/dali.py [hunk 5]
```diff
@@ -176,7 +176,11 @@ def build(config, mode='train'):
         2: types.INTERP_CUBIC,  # cv2.INTER_CUBIC
         4: types.INTERP_LANCZOS3,  # XXX use LANCZOS3 for cv2.INTER_LANCZOS4
     }
-    output_dtype = types.FLOAT16 if config.get("use_pure_fp16", False) else types.FLOAT
+   
+    if config.get("use_pure_fp16", False) and config.get("use_amp", False):
```
[the remark this answers is not in the document]
Comment (huangxu96) [reply tools/static/dali.py]: Done.

File: tools/static/program.py (hunk 3 above)
Comment (huangxu96) [reply tools/static/program.py:24]: Done.

File: ppcls/modeling/loss.py (hunk 1 above)
Comment (huangxu96) [reply ppcls/modeling/loss.py:44]: 这部分已经统一都使用mean了。

File: tools/static/program.py (hunk 4 above)
Comment (huangxu96) [reply tools/static/program.py:229]: 同上，已经都使用mean了，不需要这个参数控制

File: ppcls/modeling/architectures/resnet.py (hunk 2 above)
Comment (wzzju) [reply ppcls/modeling/architectures/resnet.py:280]: 这是一个python的context manager，在这个上下文下中组网op只会在namescope属性中加个标记（用于标识该op需要使用fp16计算kernel），即染色作用。这个是不会影响fp32训练的。

**Review by littletomatodonkey — APPROVED** [review]
